postgres/datadog_checks/postgres/postgres.py

```diff
@@ -180,10 +181,39 @@ def _new_query_executor(self, queries):
         )
 
     def execute_query_raw(self, query):
-        with self._get_main_db().cursor() as cursor:
-            cursor.execute(query)
-            rows = cursor.fetchall()
-            return rows
+        with self.db() as conn:
+            with conn.cursor() as cursor:
+                cursor.execute(query)
+                rows = cursor.fetchall()
+                return rows
+
+    @contextlib.contextmanager
+    def db(self):
+        """
+        db context manager that yields a healthy connection to the main database
+        """
+        if not self._db or self._db.closed:
+            # if the connection is closed, we need to reinitialize the connection
+            self._db = self._new_connection(self._config.dbname)
+            # once the connection is reinitialized, we need to reload the pg_settings
+            self._load_pg_settings(self._db)
+        if self._db.status != psycopg2.extensions.STATUS_READY:
+            self._db.rollback()
+        try:
+            yield self._db
+        except (psycopg2.InterfaceError, InterruptedError):
+            # if we get an interface error or an interrupted error,
+            # we gracefully close the connection
+            try:
+                self._db.close()
+            except Exception:
+                pass
+            finally:
+                self._db = None
+            raise
```

Comment:
so if we raise here, it will restart the check?

Reply:
also should we log the exception in that case?

Reply:
it will not restart but abort the the check. The next check run will re-establish the connection

Reply:
@jmeunier28 I was able to simulate connection termination from local docker by logging to postgres and issue a `pg_terminate_backend` to kill the active connection. 
The log is a bit messy but hopefully it explains what will happen after the change.
Below is the output:
During check run, the connection will error out and check will be aborted.
```
postgresdb-1  | 2023-10-05 19:35:31.920 GMT [33] FATAL:  terminating connection due to administrator command
postgresdb-1  | 2023-10-05 19:35:31.920 GMT [56] FATAL:  terminating connection due to administrator command
dd-agent      | 2023-10-05 19:35:33 UTC | CORE | DEBUG | (pkg/aggregator/demultiplexer.go:172 in sendIterableSeries) | Demultiplexer: sendIterableSeries: start sending iterable series to the serializer
dd-agent      | 2023-10-05 19:35:33 UTC | CORE | DEBUG | (pkg/aggregator/demultiplexer.go:178 in sendIterableSeries) | Demultiplexer: sendIterableSeries: stop routine
dd-agent      | 2023-10-05 19:35:33 UTC | CORE | DEBUG | (pkg/aggregator/aggregator.go:592 in sendServiceChecks) | Flushing 2 service checks to the forwarder
dd-agent      | 2023-10-05 19:35:36 UTC | CORE | DEBUG | (pkg/collector/worker/check_logger.go:43 in CheckStarted) | check:postgres | Running check...
dd-agent      | 2023-10-05 19:35:36 UTC | CORE | DEBUG | (pkg/collector/python/check.go:88 in runCheck) | Running python check postgres (version: '15.0.0', id: 'postgres:955aa2abfc6e3d03')
dd-agent      | 2023-10-05 19:35:36 UTC | CORE | ERROR | (pkg/collector/python/datadog_agent.go:129 in LogMessage) | postgres:955aa2abfc6e3d03 | (postgres.py:215) | Unhandled exception while using database connection dogshelter
dd-agent      | Traceback (most recent call last):
dd-agent      |   File "/usr/src/dbm-integrations-core/postgres/datadog_checks/postgres/postgres.py", line 203, in db
dd-agent      |     yield self._db
dd-agent      |   File "/usr/src/dbm-integrations-core/postgres/datadog_checks/postgres/version_utils.py", line 31, in get_raw_version
dd-agent      |     cursor.execute('SHOW SERVER_VERSION;')
dd-agent      | psycopg2.errors.AdminShutdown: terminating connection due to administrator command
dd-agent      | server closed the connection unexpectedly
dd-agent      |         This probably means the server terminated abnormally
dd-agent      |         before or while processing the request.
dd-agent      | 2023-10-05 19:35:36 UTC | CORE | ERROR | (pkg/collector/python/datadog_agent.go:129 in LogMessage) | postgres:955aa2abfc6e3d03 | (postgres.py:906) | Unable to collect postgres metrics.
dd-agent      | Traceback (most recent call last):
dd-agent      |   File "/usr/src/dbm-integrations-core/postgres/datadog_checks/postgres/postgres.py", line 886, in check
dd-agent      |     self.load_version()  # We don't want to cache versions between runs to capture minor updates for metadata
dd-agent      |   File "/usr/src/dbm-integrations-core/postgres/datadog_checks/postgres/postgres.py", line 360, in load_version
dd-agent      |     raw_version = self._version_utils.get_raw_version(self.db())
dd-agent      |   File "/usr/src/dbm-integrations-core/postgres/datadog_checks/postgres/version_utils.py", line 31, in get_raw_version
dd-agent      |     cursor.execute('SHOW SERVER_VERSION;')
dd-agent      | psycopg2.errors.AdminShutdown: terminating connection due to administrator command
dd-agent      | server closed the connection unexpectedly
dd-agent      |         This probably means the server terminated abnormally
dd-agent      |         before or while processing the request.
dd-agent      | 2023-10-05 19:35:36 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:135 in LogMessage) | postgres:955aa2abfc6e3d03 | (postgres.py:296) | Cleaning state
dd-agent      | 2023-10-05 19:35:36 UTC | CORE | ERROR | (pkg/collector/worker/check_logger.go:70 in Error) | check:postgres | Error running check: [{"message": "terminating connection due to administrator command\nserver closed the connection unexpectedly\n\tThis probably means the server terminated abnormally\n\tbefore or while processing the request.\n", "traceback": "Traceback (most recent call last):\n  File \"/usr/src/dbm-integrations-core/datadog_checks_base/datadog_checks/base/checks/base.py\", line 1235, in run\n    self.check(instance)\n  File \"/usr/src/dbm-integrations-core/postgres/datadog_checks/postgres/postgres.py\", line 918, in check\n    raise e\n  File \"/usr/src/dbm-integrations-core/postgres/datadog_checks/postgres/postgres.py\", line 886, in check\n    self.load_version()  # We don't want to cache versions between runs to capture minor updates for metadata\n  File \"/usr/src/dbm-integrations-core/postgres/datadog_checks/postgres/postgres.py\", line 360, in load_version\n    raw_version = self._version_utils.get_raw_version(self.db())\n  File \"/usr/src/dbm-integrations-core/postgres/datadog_checks/postgres/version_utils.py\", line 31, in get_raw_version\n    cursor.execute('SHOW SERVER_VERSION;')\npsycopg2.errors.AdminShutdown: terminating connection due to administrator command\nserver closed the connection unexpectedly\n\tThis probably means the server terminated abnormally\n\tbefore or while processing the request.\n\n"}]
dd-agent      | 2023-10-05 19:35:36 UTC | CORE | DEBUG | (pkg/collector/worker/check_logger.go:60 in CheckFinished) | check:postgres | Done running check
```

Next check run will re-establish a healthy connection. Notice the `pg_settings` query is ran only when a new connection is established. 
```
dd-agent      | 2023-10-05 19:35:51 UTC | CORE | DEBUG | (pkg/collector/worker/check_logger.go:43 in CheckStarted) | check:postgres | Running check...
dd-agent      | 2023-10-05 19:35:51 UTC | CORE | DEBUG | (pkg/collector/python/check.go:88 in runCheck) | Running python check postgres (version: '15.0.0', id: 'postgres:955aa2abfc6e3d03')
dd-agent      | 2023-10-05 19:35:51 UTC | CORE | DEBUG | (pkg/collector/python/datadog_agent.go:135 in LogMessage) | postgres:955aa2abfc6e3d03 | (postgres.py:714) | Running query [SELECT name, setting FROM pg_settings WHERE name IN (%s, %s, %s)]
dd-agent      | 2023-10-05 19:35:51 UTC | CORE | DEBUG | (pkg/metadata/inventories/inventories.go:194 in SetCheckMetadata) | setting check metadata for check postgres:955aa2abfc6e3d03, 'version.major': '12'
dd-agent      | 2023-10-05 19:35:51 UTC | CORE | DEBUG | (pkg/metadata/inventories/inventories.go:194 in SetCheckMetadata) | setting check metadata for check postgres:955aa2abfc6e3d03, 'version.minor': '16'
dd-agent      | 2023-10-05 19:35:51 UTC | CORE | DEBUG | (pkg/metadata/inventories/inventories.go:194 in SetCheckMetadata) | setting check metadata for check postgres:955aa2abfc6e3d03, 'version.patch': '0'
dd-agent      | 2023-10-05 19:35:51 UTC | CORE | DEBUG | (pkg/metadata/inventories/inventories.go:194 in SetCheckMetadata) | setting check metadata for check postgres:955aa2abfc6e3d03, 'version.raw': '12.16 (Debian 12.16-1.pgdg120+1)'
```

Reply:
excellent, that looks great 👍